include/circt/Dialect/Moore/MooreOps.td

```diff
@@ -958,4 +958,69 @@ def UnionCreateOp : MooreOp<"union_create"> {
   type($input) `->` type($result)
   }];
  }
+
+def ConditionalOp : MooreOp<"conditional",[
+  RecursiveMemoryEffects,
+  NoRegionArguments, 
+  SingleBlockImplicitTerminator<"moore::YieldOp">
+]> {
+  let summary = "Conditional operation";
+  let description = [{
+    If cond_predicate is true, the operator returns the value of the first 
+    expression without evaluating the second expression; if false, it returns 
+    the value of the second expression without evaluating the first expression.
+    If cond_predicate evaluates to an ambiguous value (x or z), then both the
+    first expression and the second expression shall be evaluated, and compared
+    for logical equivalence. If that comparison is true (1), the operator shall
+    return either the first or second expression. Otherwise the operator returns
+    a result based on the data types of the expressions.
+
+    When both the first and second expressions are of integral types, if the
+    cond_predicate evaluates to an ambiguous value and the expressions are not
+    logically equivalent, their results shall be combined bit by bit using the
+    table below to calculate the final result. The first and second expressions
+    are extended to the same width.
+
+    |?: | 0 | 1 | X | Z |
+    |---|---|---|---|---|
+    | 0 | 0 | X | X | X |
+    | 1 | X | 1 | X | X |
+    | X | X | X | X | X |
+    | Z | X | X | X | X |
+
+    See IEEE 1800-2017 § 11.4.11 "Conditional operator".
+  }];
+  let arguments = (ins AnySingleBitType:$condition);
+  let results = (outs UnpackedType:$result);
+  let regions = (region SizedRegion<1>:$trueRegion,
+                        SizedRegion<1>:$falseRegion);
+  let assemblyFormat = [{
+    $condition attr-dict `:` type($condition) `->` type($result)
+    $trueRegion $falseRegion
+  }];
+}
+
+def YieldOp : MooreOp<"yield", [
+  Pure, 
+  Terminator, 
+  HasParent<"ConditionalOp">
+]> {
+  let summary = "conditional yield and termination operation";
+  let description = [{
+    "moore.yield" yields an SSA value from the Moore dialect op region and
+    terminates the regions. The semantics of how the values are yielded is
+    defined by the parent operation.
+    If "moore.yield" has any operands, the operands must match the parent
+    operation's results.
```

Comment:
How do you ensure the types of operands of `moore.yield` match the result types of `moore.conditional`? Maybe you should add the related method to check this.

Reply:
Nice advice! However, Slang already ensures a match between result type and expressions type, so additional checks seem unnecessary.

Reply:
I think you still need a check to make sure that the IR is not malformed. Most of the time the IR will come from import through Slang, but in many other cases the input IR will be handwritten in case of unit tests, or it will be generated from transformation passes. It is important that operations check that the IR is correct to catch bugs in those other cases as soon as possible.

The `hw::OutputOp`, `arc::OutputOp`, `scf::YieldOp`, or `func::ReturnOp` have to do a similar kind of check -- you might find some inspiration on how to do it there. I'm not 100% if the output/yield op verifies that it matches the parent condition/module/arc/func/if op, or if the parent op checks that it has the right terminator.